File: samples/nodejs-typescript-food-catalog/src/functions/http-notification.ts
import { app, HttpRequest, InvocationContext } from "@azure/functions";
import { ConnectionMessage } from "../common/ConnectionMessage";
import { getQueueClient } from "../common/queueClient";
import { validateToken } from "../common/validateToken";
import { streamToJson } from "../common/utils";
import { config } from "../common/config";

enum TargetConnectorState {
    Enabled = 'enabled',
    Disabled = 'disabled',
}

app.http('notification', {
    methods: ['POST'],
    handler: async (request: HttpRequest, context: InvocationContext) => {
        const body = await streamToJson(request.body);
        context.log('Received notification');
        context.log(JSON.stringify(body, null, 2));

        const {
            aadAppTenantId: tenantId,
            aadAppClientId: clientId
        } = config;

        const token = body?.validationTokens[0];
        context.log(`Validating token: ${token}, tenantId: ${tenantId}, clientId: ${clientId}...`);
        await validateToken(token, tenantId, clientId);
        context.log('Token validated');

        const changeDetails = body?.value[0]?.resourceData;
        const targetConnectorState = changeDetails?.state;

        const message: ConnectionMessage = {
            connectorId: changeDetails?.id,
            connectorTicket: changeDetails?.connectorsTicket
        }

        if (targetConnectorState === TargetConnectorState.Enabled) {
            message.action = 'create';
        }
        else if (targetConnectorState === TargetConnectorState.Disabled) {
            message.action = 'delete';
        }

        if (!message.action) {
            context.error('Invalid action');
            return;
        }

        context.log(JSON.stringify(message, null, 2));

        const queueClient = await getQueueClient('queue-connection');
        const messageString = btoa(JSON.stringify(message));
        context.log('Sending message to queue queue-connection: ${message}');
        // must base64 encode
        await queueClient.sendMessage(messageString);
        context.log('Message sent');

        return {
            status: 202
        }
    }
})

File: samples/dotnet-csharp-graphdocs-ttk/Functions/HttpNotification.cs
using Azure.Storage.Queues;
using GraphDocsConnector.Messages;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using System.Diagnostics;
using System.Text.Json;

namespace GraphDocsConnector.Functions
{
    public class HttpNotification
    {
        private readonly ILogger<HttpNotification> _logger;
        private readonly QueueClient _queueClient;
        private readonly IConfiguration _configuration;

        public HttpNotification(QueueServiceClient queueClient, ILogger<HttpNotification> logger, IConfiguration configuration)
        {
            _queueClient = queueClient.GetQueueClient("queue-connection");
            _logger = logger;
            _configuration = configuration;
        }

        [Function("notification")]
        public async Task<IActionResult> Run([HttpTrigger(AuthorizationLevel.Anonymous, "post")] HttpRequest req)
        {
            var notification = await req.ReadFromJsonAsync<NotificationMessage>();
            await ProcessNotification(notification);
            return new AcceptedResult();
        }

        private async Task ProcessNotification(NotificationMessage? notificationMessage)
        {
            if (notificationMessage is null)
            {
                _logger.LogWarning("Notification message is null");
                return;
            }

            var tenantId = _configuration["Entra:TenantId"];
            var clientId = _configuration["Entra:ClientId"];
            Debug.Assert(tenantId is not null);
            Debug.Assert(clientId is not null);

            var tokenValidator = new TokenValidator(tenantId, clientId);
            if (notificationMessage.ValidationTokens is null ||
                !notificationMessage.ValidationTokens.Any())
            {
                _logger.LogError("No token to validate found");
                return;
            }
            
            if (!tokenValidator.ValidateToken(notificationMessage.ValidationTokens.First()))
            {
                _logger.LogError("Invalid token");
                return;
            }

            var changeDetails = notificationMessage.Value?[0].ResourceData;
            var targetConnectionState = changeDetails?.State;

            if (changeDetails is null || targetConnectionState is null)
            {
                _logger.LogError("Notification payload doesn't contain the necessary information");
                return;
            }

            var message = new ConnectionMessage
            {
                ConnectorId = changeDetails.Id,
                ConnectorTicket = changeDetails.ConnectorsTicket
            };

            switch (targetConnectionState)
            {
                case "enabled":
                    message.Action = ConnectionMessageAction.Create;
                    break;
                case "disabled":
                    message.Action = ConnectionMessageAction.Delete;
                    break;
                default:
                    return;
            }

            await _queueClient.CreateIfNotExistsAsync();
            await _queueClient.SendMessageAsync(JsonSerializer.Serialize(message));
        }
    }
}
